Remove Generic<T> Exception code repetition

benaadams opened the pull request:
With the introduction of https://github.com/dotnet/coreclr/pull/6103 "Do not inline methods that never return"

These changes reduce the code gen size of the jitted methods (regular and generic * N) for Array and ArraySegment. (And a bunch of other generic types)

Also results in a 55kB reduction int the size of System.Private.CoreLib native (Windows_NT.x64.Release).
```
Pre  System.Private.CoreLib.dll 2,346,496 bytes
Post System.Private.CoreLib.dll 2,335,744 bytes
Pre  System.Private.CoreLib.ni.dll 10,913,280 bytes
Post System.Private.CoreLib.ni.dll 10,857,472 bytes
```

55,808 byte reduction


```
Total bytes of diff: -44551 (-1.44 % of base)
    diff is an improvement.

Total byte diff includes 562 bytes from reconciling methods
        Base had    1 unique methods,       76 unique bytes
        Diff had    9 unique methods,      638 unique bytes

Top file improvements by size (bytes):
      -44551 : System.Private.CoreLib.dasm (-1.44 % of base)

1 total files with size differences.

Top method regessions by size (bytes):
         418 : System.Private.CoreLib.dasm - CustomAttributeData:.ctor(ref,struct):this
         289 : System.Private.CoreLib.dasm - CustomAttributeData:Init(ref):this
         234 : System.Private.CoreLib.dasm - TimeSpan:TryParseExact(ref,ref,ref,byref):bool
         229 : System.Private.CoreLib.dasm - DateTimeParse:ParseByFormat(byref,byref,byref,ref,byref):bool
         187 : System.Private.CoreLib.dasm - DateTimeFormat:FormatCustomized(struct,ref,ref,struct):ref

Top method improvements by size (bytes):
       -3336 : System.Private.CoreLib.dasm - Array:Sort(ref,int,int,ref)
       -2396 : System.Private.CoreLib.dasm - Array:LastIndexOf(ref,struct,int,int):int
       -1677 : System.Private.CoreLib.dasm - Array:IndexOf(ref,struct,int,int):int
       -1251 : System.Private.CoreLib.dasm - Array:BinarySearch(ref,int,int,struct,ref):int
        -884 : System.Private.CoreLib.dasm - Comparer`1:Create(ref):ref

389 total methods with size differences.
```

One of the big wins, other than the [included throw code is smaller (21 bytes of il to 7 bytes)](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/ThrowHelper.cs#L7-L36), is it prevents all the inlining of `Environment:GetResourceString(ref):ref` e.g.

Pre
```
Inlines into 060027E0 ArraySegment`1:.ctor(ref,int,int):this
  [0 IL=0008 TR=000119 06002512] [FAILED: unprofitable inline] ArgumentNullException:.ctor(ref):this
  [1 IL=0028 TR=000091 060034F5] [below ALWAYS_INLINE size] Environment:GetResourceString(ref):ref
    [0 IL=0001 TR=000128 060034F4] [FAILED: unprofitable inline] Environment:GetResourceStringLocal(ref):ref
  [0 IL=0033 TR=000102 06002AEE] [FAILED: unprofitable inline] ArgumentOutOfRangeException:.ctor(ref,ref):this
  [2 IL=0053 TR=000069 060034F5] [below ALWAYS_INLINE size] Environment:GetResourceString(ref):ref
    [0 IL=0001 TR=000136 060034F4] [FAILED: unprofitable inline] Environment:GetResourceStringLocal(ref):ref
  [0 IL=0058 TR=000080 06002AEE] [FAILED: unprofitable inline] ArgumentOutOfRangeException:.ctor(ref,ref):this
  [3 IL=0077 TR=000048 060034F5] [below ALWAYS_INLINE size] Environment:GetResourceString(ref):ref
    [0 IL=0001 TR=000144 060034F4] [FAILED: unprofitable inline] Environment:GetResourceStringLocal(ref):ref
  [0 IL=0082 TR=000059 06002E5A] [FAILED: unprofitable inline] ArgumentException:.ctor(ref):this
Budget: initialTime=390, finalTime=390, initialBudget=3900, currentBudget=3900
Budget: initialSize=2639, finalSize=2639
```
Post
```
Inlines into 06002A5B ArraySegment`1:.ctor(ref,int,int):this
  [0 IL=0004 TR=000067 06000C6F] [FAILED: does not return] ThrowHelper:ThrowArgumentNullException(int)
  [0 IL=0016 TR=000061 06000C74] [FAILED: noinline per IL/cached result] ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
  [0 IL=0028 TR=000054 06000C74] [FAILED: noinline per IL/cached result] ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
  [0 IL=0043 TR=000048 06000C6D] [FAILED: noinline per IL/cached result] ThrowHelper:ThrowArgumentException(int)
Budget: initialTime=270, finalTime=270, initialBudget=2700, currentBudget=2700
Budget: initialSize=1727, finalSize=1727
```
Which will be repeated for each `new Generic<struct>` or `Method<struct>(...)` flavour.

benaadams commented:
Just rebuilding the without changes version as that seems high

AndyAyersMS commented on src/mscorlib/src/System/Array.cs:
```diff
@@ -2636,6 +2636,47 @@ public Object Clone()
+            throw new ArgumentNullException("elementType");
+        }
+
+        private static void ThrowArgumenCountOutOfRange()
```
Typo `Argumen`

benaadams replied:
Fixed

AndyAyersMS commented on src/mscorlib/src/System/Array.cs:
```diff
@@ -2636,6 +2636,47 @@ public Object Clone()
+            throw new ArgumentOutOfRangeException("count", Environment.GetResourceString("ArgumentOutOfRange_Count"));
+        }
+
+        private static void ThrowArgumenStartIndexOutOfRange()
```
Ditto

benaadams replied:
Fixed

benaadams commented:
Assume fixing a typo didn't break the tests O_o

@dotnet-bot retest OSX x64 Checked Build and Test
@dotnet-bot retest Ubuntu x64 Checked Build and Test
@dotnet-bot retest Windows_NT x64 Debug Build and Test

justinvp commented on src/mscorlib/src/System/Array.cs:
```diff
@@ -2636,6 +2636,47 @@ public Object Clone()
+            throw new ArgumentException(Environment.GetResourceString("Arg_MustBeType"), "elementType");
+        }
+
+        private static void ThrowArgumentOutOfRangeNeedNonNegNum(string parmName)
```
Nit: `parmName` -> `paramName`

benaadams replied:
Fixed

justinvp commented on src/mscorlib/src/System/Array.cs:
```diff
@@ -2636,6 +2636,47 @@ public Object Clone()
+            throw new ArgumentOutOfRangeException(parmName, Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
+        }
+
+        private static void ThrowArgumentOutOfRangeHugeArrayNotSupported(string parmName)
```
Nit: `parmName` -> `paramName`

benaadams replied:
Fixed

AndyAyersMS commented:
Unfortunately this change breaks the way that `Internal:CommonlyUsedGenericInstantiations` is used to force selected generic instantiations during the crossgen of the core library.
```C#
        static void CommonlyUsedGenericInstantiations()
        {
            System.Array.Sort<double>(null);
            System.Array.Sort<int>(null);
            System.Array.Sort<IntPtr>(null);
            new ArraySegment<byte>(new byte[1], 0, 0);
           ...
```
resulting asm with this change:
```asm
_M1895_IG01:
       sub      rsp, 120
       xor      rax, rax
       mov      qword ptr [rsp+68H], rax
       mov      qword ptr [rsp+70H], rax
G_M1895_IG02:
       call     Array:ThrowArrayNullException()
       int3
```
The jit now can inline the first `Sort` call and propagate the null and realize that the method will throw an exception, and so the jit dead codes the remainder of the method. While this optimization is "correct", the upshot is the referenced generic instantiations do not get created during crossgen. I believe we can just mark this method as no optimization and get back to the desired behavior and protect ourselves from future jit cleverness. Fixing this will cut down the size savings from this change, perhaps by 50% or so...

benaadams commented:
Not sure I understand, wouldn't the previous behaviour have the same effect?
```csharp
public static void Sort<T>(T[] array)
{
    if (array == null)
        throw new ArgumentNullException("array");
    Contract.EndContractBlock();
    Sort<T>(array, 0, array.Length, null);
}
```
Should I back out the `ThrowArrayNullException()` call for this method?

benaadams commented:
Also is it worth adding some Array.Empty to that list? e.g.
```csharp
System.Array.Empty<byte>();
System.Array.Empty<char>();
System.Array.Empty<int>();
System.Array.Empty<uint>();
System.Array.Empty<long>();
System.Array.Empty<ulong>();
System.Array.Empty<object>();
```

benaadams commented:
Best of both worlds?
```csharp
Array.Sort<double>(Array.Empty<double>());
Array.Sort<int>(Array.Empty<int>());
Array.Sort<IntPtr>(Array.Empty<IntPtr>());
```

benaadams commented:
Adding the `Array.Empty()` Pushes it back up to the pre-size on disk 10,913,280 bytes

benaadams commented:
Removed some more repeated exceptions to functions specically in Search<T> path, down to 10,889,216 bytes

So saving 24064 bytes on disk or 24kB

mikedn commented on src/mscorlib/src/System/Array.cs:
```diff
@@ -2636,6 +2636,47 @@ public Object Clone()
         public extern void Initialize();
+
+
+        private static void ThrowArrayNullException()
```
You should look at the existing `ThrowHelper` class and use that whenever possible. It's not very useful if we keep adding such methods all other the place.

benaadams replied:
Changed all the calls to ThrowHelper rather than the statics

benaadams commented:
Should be able to push a lot of this to ThrowHelper

mikedn commented on src/mscorlib/src/System/Array.cs:
```diff
@@ -45,7 +45,7 @@ public abstract class Array : ICloneable, IList, IStructuralComparable, IStructu
         public static void Resize<T>(ref T[] array, int newSize) {
             if (newSize < 0)
-                throw new ArgumentOutOfRangeException("newSize", Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
+                ThrowArgumentOutOfRangeNeedNonNegNum(nameof(newSize));
```
Use of constant strings results in an additional call, that's why ThrowHelper passes the argument as an enum instead.

benaadams replied:
The diffs are a bit easier to compare with ThrowHelper also :smile:

benaadams commented:
Down to 10,887,168 bytes on disk with ThrowHelper saving 26112 bytes or 25kB

Also now none of the generic methods throw directly; for jit time

benaadams commented:
Down to 10,885,632 bytes with ArraySegment; 27,648 byte saving

benaadams commented:
@dotnet-bot retest Windows_NT x64 Release Priority 1 Build and Test

AndyAyersMS commented:
Circling back a bit to my note on the change to codegen for `CommonlyUsedGenericInstantiations`:  The refactoring you did had the side effect of making the array sort methods look like good inline candidates. Previously they weren't. So it uncovered an implicit dependence on the jit's behavior. The mechanism is fragile; we could just as easily have run into this in other ways
```
BEFORE

Inlines into 06000BE1 Internal:CommonlyUsedGenericInstantiations()
  [0 IL=0001 TR=000002 06002B6B] [FAILED: unprofitable inline] Array:Sort(ref)
  [0 IL=0007 TR=000006 06002B6B] [FAILED: unprofitable inline] Array:Sort(ref)
  [0 IL=0013 TR=000010 06002B6B] [FAILED: unprofitable inline] Array:Sort(ref)
  [0 IL=0026 TR=000029 060027E0] [FAILED: too many il bytes] ArraySegment`1:.ctor(ref,int,int):this
  ...

AFTER

Inlines into 06000BE1 Internal:CommonlyUsedGenericInstantiations()
  [1 IL=0001 TR=000002 06002B6B] [profitable inline] Array:Sort(ref)
    [0 IL=0003 TR=000746 06002B76] [FAILED: does not return] Array:ThrowArrayNullException()
    [0 IL=0014 TR=000755 06002B71] [FAILED: too many il bytes] Array:Sort(ref,int,int,ref)
  [2 IL=0007 TR=000006 06002B6B] [profitable inline] Array:Sort(ref)
    [0 IL=0003 TR=000767 06002B76] [FAILED: does not return] Array:ThrowArrayNullException()
    [0 IL=0014 TR=000776 06002B71] [FAILED: too many il bytes] Array:Sort(ref,int,int,ref)
  [3 IL=0013 TR=000010 06002B6B] [profitable inline] Array:Sort(ref)
    [0 IL=0003 TR=000788 06002B76] [FAILED: does not return] Array:ThrowArrayNullException()
    [0 IL=0014 TR=000797 06002B71] [FAILED: too many il bytes] Array:Sort(ref,int,int,ref)
  [0 IL=0026 TR=000029 060027E0] [FAILED: too many il bytes] ArraySegment`1:.ctor(ref,int,int):this
 ...
```

benaadams commented:
Would that suggest it would be sensible to go straight for `Array:Sort(ref,int,int,ref)` in `Internal:CommonlyUsedGenericInstantiations` and leave `Array:Sort(ref)` to be inlined in user code?

benaadams commented:
It also drops it a few more bytes to 10,884,608 bytes (1024 more?)

AndyAyersMS commented:
I believe the intent of `CommonlyUsedGenericInstantiations` is that for generic methods the most commonly used method should be the one that is instantiated. This will transitively cause instantiations of the methods it invokes.

The top-level Sort calls are getting inlined. @jkotas can correct me here but I believe that means these Sort methods will not get instantiated as expected, since the instantiation closure process is driven by "fixup" references left in the final code stream.

So my preference would be mark `CommonlyUsedGenericInstantiations` with `[MethodImplAttribute(MethodImplOptions.NoOptimization)]`, along with a comment indicating why we don't want the jit to optimize this method, and revert it back to calling the simpler Sort methods that were there initially.

I am not sure if the empty array variants are used often enough to warrant forced instantiation, but likely it is fairly harmless to add them.

The change to consistently use `ThrowHelper` is a change for the better; however it will sometimes inhibit laying out the code so that the throwing methods are moved to the end of the method, since the jit has no way of transitively deducing the no return property.  There are also cases (both before and after your changes) where the initial part of the `ThrowHelper` call chain gets inlined; this seems somewhat undesirable, to the point where I'd consider marking all the entry point methods in the `ThrowHelper` class as no inline. However doing that would inhibit even more cases where we get the beneficial layout.

Sorting all this out is going to take some work, so for now  I'd just fix the issue with `CommonlyUsedGenericInstantiations`.

You can see the range of behaviors on display in `Array:CreateInstance(ref,ref):ref` after your change:
```asm
G_M62283_IG01:
       push     rdi
       push     rsi
       push     rbp
       push     rbx
       sub      rsp, 40
       xor      rax, rax
       mov      qword ptr [rsp+20H], rax
       mov      rsi, rdx
G_M62283_IG02:
       test     rcx, rcx
       je       G_M62283_IG16
G_M62283_IG03:
       test     rsi, rsi
       je       G_M62283_IG17
G_M62283_IG04:
       mov      edi, dword ptr [rsi+8]
       test     edi, edi
       je       G_M62283_IG18
G_M62283_IG05:
       mov      rax, qword ptr [rcx]
       mov      rax, qword ptr [rax+152]
       call     gword ptr [rax+40]Type:get_UnderlyingSystemType():ref:this
       mov      rdx, rax
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_ISINSTANCEOFCLASS
       mov      rbx, rax
       test     rbx, rbx
       jne      SHORT G_M62283_IG06
       mov      ecx, 49
       mov      edx, 27
       call     ThrowHelper:ThrowArgumentException(int,int)     // not deduced no-return
G_M62283_IG06:
       xor      ebp, ebp
       test     edi, edi
       jle      SHORT G_M62283_IG09
G_M62283_IG07:
       movsxd   rcx, ebp
       cmp      dword ptr [rsi+4*rcx+16], 0
       jl       G_M62283_IG19
G_M62283_IG08:
       inc      ebp
       cmp      edi, ebp
       jg       SHORT G_M62283_IG07
G_M62283_IG09:
       test     edi, edi
       jne      SHORT G_M62283_IG11
G_M62283_IG10:
       xor      rcx, rcx
       mov      bword ptr [rsp+20H], rcx
       jmp      SHORT G_M62283_IG12
G_M62283_IG11:
       cmp      edi, 0
       jbe      G_M62283_IG20
       add      rsi, 16
       mov      bword ptr [rsp+20H], rsi
G_M62283_IG12:
       mov      rcx, rbx
       mov      rax, qword ptr [rbx]
       mov      rax, qword ptr [rax+88]
       call     gword ptr [rax+24]Type:get_TypeHandle():struct:this
       test     rax, rax
       jne      SHORT G_M62283_IG13
       xor      rcx, rcx
       jmp      SHORT G_M62283_IG14
G_M62283_IG13:
       mov      rcx, qword ptr [rax+24]
G_M62283_IG14:
       mov      edx, edi
       mov      r8, bword ptr [rsp+20H]
       xor      r9, r9
       call     Array:InternalCreate(long,int,long,long):ref
       nop      
G_M62283_IG15:
       add      rsp, 40
       pop      rbx
       pop      rbp
       pop      rsi
       pop      rdi
       ret      
************** Beginning of cold code **************
G_M62283_IG16:
       mov      ecx, 27
       call     ThrowHelper:ThrowArgumentNullException(int)    // moved, deduced no return
G_M62283_IG17:
       mov      ecx, 34
       call     ThrowHelper:ThrowArgumentNullException(int)   // moved, deduced no return
G_M62283_IG18:
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_NEWSFAST
       mov      rbp, rax
       mov      ecx, 0x10072
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      rcx, rax
       call     Environment:GetResourceStringLocal(ref):ref
       mov      rdx, rax
       mov      rcx, rbp
       call     ArgumentException:.ctor(ref):this                        // moved,  inlined
       mov      rcx, rbp
       call     CORINFO_HELP_THROW
G_M62283_IG19:
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       mov      ecx, 0x1009C
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      rdi, rax
       mov      dword ptr [rsi+8], ebp
       mov      ecx, 847
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      r8, rax
       mov      rcx, rdi
       mov      rdx, rsi
       call     String:Concat(ref,ref,ref):ref
       mov      rsi, rax
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_NEWSFAST
       mov      rdi, rax
       mov      ecx, 0x4D2F
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      rcx, rax
       call     Environment:GetResourceStringLocal(ref):ref
       mov      rbx, rax
       mov      rcx, rdi
       call     Exception:Init():this
       lea      rcx, bword ptr [rdi+32]
       mov      rdx, rbx
       call     CORINFO_HELP_ASSIGN_REF
       mov      dword ptr [rdi+132], 0xD1FFAB1E
       lea      rcx, bword ptr [rdi+144]
       mov      rdx, rsi
       call     CORINFO_HELP_ASSIGN_REF
       mov      dword ptr [rdi+132], 0xD1FFAB1E
       mov      dword ptr [rdi+132], 0xD1FFAB1E
       mov      rcx, rdi
       call     CORINFO_HELP_THROW                                 //  moved, inlined
       int3     
G_M62283_IG20:
       call     CORINFO_HELP_RNGCHKFAIL
       int3 
```

benaadams commented:
Marked as such.

Up a bit to 10,885,120 bytes so 28,160 bytes saving

benaadams commented:
@AndyAyersMS thank you for working though this with me. Are the various outputs from jitutils? (dasm, and inlining choices)

benaadams commented:
There are a couple throws I haven't moved out to the ThrowHelper in non-generic methods (for example in CreateInstance). Shall I do those also?

There are a couple oddities:
```csharp
ThrowHelper.ThrowArgumentOutOfRangeException(
    length1 < 0 ? ExceptionArgument.length1 : ExceptionArgument.length2, 
    ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
```
Which I could split two 2 ifs?

It would only leave the 3 which have variable parameter names e.g.
```csharp
throw new ArgumentOutOfRangeException("lengths[" + i + ']', 
    Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
```
Which I could move out to a static
```csharp
static void ThrowLengthOutOfRange(int parameterNumber)
{
    throw new ArgumentOutOfRangeException("lengths[" + parameterNumber + ']', 
        Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
}
```
Then see how it looks?

AndyAyersMS commented:
Here's what I have been doing to get at the details. It's a bit clunky but doable.

**DIFFS**

First, have two CoreCLR repos synced at the same master commit (call these BASE and DIFF). In DIFF, pull in your changes, switch to that branch, and rebase the changes onto master. Build both REL and CHK in each repo (skipping test builds). In each repo, copy the CHK clrjit.dll over the REL clrjit.dll:
```
copy c:\repos\BASE\bin\Product\Windows_NT.x64.Checked\clrjit.dll  
         c:\repos\BASE\bin\Product\Windows_NT.x64.release

(ditto for DIFF)
```
Run jit-diffs on each repo, something like:
```
jit-diff diff -t DIFF -b c:\repos\BASE\bin\Product\Windows_NT.x64.release\crossgen.exe -c --core_root c:\repos\BASE\bin\Product\Windows_NT.x64.release

jit-diff diff -t DIFF -b c:\repos\DIFF\bin\Product\Windows_NT.x64.release\crossgen.exe -c --core_root c:\repos\DIFF\bin\Product\Windows_NT.x64.release
```
This is a bit inefficient since each jit-diff run runs crossgen 2x. Depending on how you have jit-diff configured the "diff" crossgen leg these runs may fail. You can ignore that since we only care about the base runs (there's probably a more streamlined way to do this using the underlying tool jit-dasm directly, but the above is fast enough I haven't bothered to figure it out).

In my setup, jit-diff writes the disassembly files to directories under c:\asmdiff.

Now analyze:
```
jit-analyze --base c:\asmdiffs\BASE\base --diff c:\asmdiffs\DIFF\base
```
To look at particular method diffs, you can run some file diff tool on the .dasm files in the two directories. In this case there will be a lot of spurious diffs -- for instance string literal indices have changed -- so using a file comparer that can exclude diffs based on regular expressions may come in handy. I've had good experiences with BeyondCompare.

**INLINES**

Since we've swapped a CHK jit into the REL directores, we can run the crossgens in each repo with suitable COMPlus settings to get detailed jit outputs. For instance:
```
set COMPlus_JitPrintInlinedMethods=1
cd c:\repos\BASE\bin\Product\Windows_NT.x64.release
crossgen System.Private.CoreLib.dll > c:\asmdiffs\base-inl.log

(ditto for DIFF)
```
Then compare the two logs to see the inlining changes.

AndyAyersMS commented:
As far as further changes -- from the caller's standpoint the less fussing about the better. Deferring formatting and such to a `ThrowHelper` callee seems sensible, especially so if there are multiple call sites.

For the `if` case I'd do likewise, pass length into the helper and let it decide how to proceed.

benaadams commented:
Moved all exceptions to ThrowHelper and split the tenary exceptions into two, as the ThrowHelper is absorbing the cost of the throw.

Hopefully should be looking better.

AndyAyersMS commented:
LGTM. Somebody from the runtime/framework side should weigh in too before we take this. 

@jkotas can you take a look?

jamesqo commented on src/mscorlib/src/System/ThrowHelper.cs:
```diff
@@ -227,6 +248,94 @@ internal static class ThrowHelper {
+                    argumentName = "newSize";
+                    break;
+
+                case ExceptionArgument.len:
```
Maybe all of this could use `Enum.ToString` instead, since it's not a performance-critical codepath?

jasonwilliams200OK replied:
Yeah, the two adjacent methods can probably be replaced by:

```c#
internal static string GetArgumentName(ExceptionArgument argument)
{
    if(!Enum.IsDefined(typeof(ExceptionArgument), argument))
    {
        Contract.Assert(false, "The enum value is not defined, please check the ExceptionArgumentName Enum.");
        return string.Empty;
    }

    return argument.ToString();
}

internal static string GetResourceName(ExceptionResource resource)
{
    if(!Enum.IsDefined(typeof(ExceptionResource), resource))
    {
        Contract.Assert( false, "The enum value is not defined, please check the ExceptionArgumentName Enum.");
        return string.Empty;
    }

    return resource.ToString();
}
```

benaadams replied:
Looks like it should work; Enum.ToString() looks like it could do with some improves - (though that's a different PR)

benaadams replied:
I assume the `Enum.ToString()` won't do anything unusual in AOT, locales etc

benaadams replied:
Looks like already a PR for `Enum.ToString()` https://github.com/dotnet/coreclr/pull/3849

mikedn commented:
@AndyAyersMS 
```
mov      ecx, 49
mov      edx, 27
call     ThrowHelper:ThrowArgumentException(int,int)     // not deduced no-return
```

I'm not sure what happens here. Presumably inlining was rejected before the no-return info was computed? I'll take a look.

benaadams commented:
Changed to use Enum.ToString for ThrowHelper, less chance of typing error as previously 3 strings had to match.

Down to 10,877,440 bytes on disk

jasonwilliams200OK commented:
@benaadams, if it makes sense, can you add the comparison for System.Private.CoreLib.dll (non-native / MSIL) version in description as well?

benaadams commented:
@jasonwilliams200OK added, so its a 4096 byte change in il for a 35,840 byte change in pre-compilation; and more at jit time.

benaadams commented:
@dotnet-bot retest OSX x64 Checked Build and Test

benaadams commented:
Squashed to two commits

benaadams commented:
@dotnet-bot retest Windows_NT x64 Debug Build and Test

justinvp commented on src/mscorlib/src/System/ThrowHelper.cs:
```diff
@@ -85,18 +85,39 @@ internal static class ThrowHelper {
 
+        internal static void ThrowArgumentOutOfRangeException(ExceptionArgument argument, int paramNumber, ExceptionResource resource)
+        {
+            throw new ArgumentOutOfRangeException(GetArgumentName(argument) + "[" + paramNumber + ']',
```
I realize this is an error path, so extra allocations aren't as much of a concern, but the `paramName` could be passed to `ArgumentOutOfRangeException` as:

```c#
GetArgumentName(argument) + "[" + paramNumber.ToString() + "]"
```

 - `paramNumber` -> `paramNumber.ToString()` avoids a box allocation and avoids the `object[]` array allocation with the call to `string.Concat(object[])` (`string.Concat(string,string,string,string)` will be used instead).
 - `']'` -> `"]"` isn't strictly necessary as Roslyn is smart enough to treat this as a string constant, instead of as a char that will box and cause `string.Concat(object[])` to be called, but is for consistency with earlier `"["`.

benaadams replied:
Changed and squashed into first commit

justinvp replied:
Super nit picky: Looks like there is a misplaced space char before the `.ToString()` and there should be a space between the `ToString()` and `+`, e.g.:

Currently:
 
```c#
paramNumber .ToString()+ "]"
```

Should be:

```csharp
paramNumber.ToString() + "]"
```

jkotas commented:
This outlining optimization would be best done by IL-to-IL optimizer, without changing code everywhere. For this particular case, it can go even further depending on what you are optimizing for, e.g. if you are producing single .exe optimized for size and do not care about good error messages, the IL-to-IL optimization can strip them; or even turn all argument validation via exceptions into debug breaks or fail fasts.

We do have this outlining optimization in the internal .NET Native for UWP toolchain, but it is unfortunately written in way that does not make it easy to package it as .NET Core-friendly part. It is also the reason why you won't see the ThrowHelper used in CoreRT or CoreFX - it throws off the outlining optimization.

Until the IL to IL optimizer emerges, I would be ok with doing it manually in CoreCLR corelib where it has significant benefit as stop-gap measure, like for frequently used generic types because of the effect is multiplied for them - existing List<T> or Dictionary<T> uses are good example. But I do not think it makes sense to do it manually across the board.

BTW: Number of optimizations would be a good fit for standalone IL to IL optimizer. For example, Linq query optimizations (#2385), or stripping unused private implementation details - if you do not need to reflect over them. Similar optimizations have been historically bundled into IL obfuscators.

benaadams commented:
List and Dictionary were already using ThrowHelper, I think its where it came from. I've covered all the hot generics and a couple extra things accidentally.

System.Private.CoreLib.dll 2,335,744 bytes, 10,752 byte reduction
System.Private.CoreLib.ni.dll 10,857,472 bytes, 55,808 byte reduction

Let me know if you want anything backed out?

benaadams commented:
I got stuck in a rhythm, I think... O_o

benaadams commented:
@AndyAyersMS got jit-diff working - thank you :grinning: 
```
Total bytes of diff: -44551 (-1.44 % of base)
    diff is an improvement.

Total byte diff includes 562 bytes from reconciling methods
        Base had    1 unique methods,       76 unique bytes
        Diff had    9 unique methods,      638 unique bytes

Top file improvements by size (bytes):
      -44551 : System.Private.CoreLib.dasm (-1.44 % of base)

1 total files with size differences.

Top method regessions by size (bytes):
         418 : System.Private.CoreLib.dasm - CustomAttributeData:.ctor(ref,struct):this
         289 : System.Private.CoreLib.dasm - CustomAttributeData:Init(ref):this
         234 : System.Private.CoreLib.dasm - TimeSpan:TryParseExact(ref,ref,ref,byref):bool
         229 : System.Private.CoreLib.dasm - DateTimeParse:ParseByFormat(byref,byref,byref,ref,byref):bool
         187 : System.Private.CoreLib.dasm - DateTimeFormat:FormatCustomized(struct,ref,ref,struct):ref

Top method improvements by size (bytes):
       -3336 : System.Private.CoreLib.dasm - Array:Sort(ref,int,int,ref)
       -2396 : System.Private.CoreLib.dasm - Array:LastIndexOf(ref,struct,int,int):int
       -1677 : System.Private.CoreLib.dasm - Array:IndexOf(ref,struct,int,int):int
       -1251 : System.Private.CoreLib.dasm - Array:BinarySearch(ref,int,int,struct,ref):int
        -884 : System.Private.CoreLib.dasm - Comparer`1:Create(ref):ref

389 total methods with size differences.
```

benaadams commented:
One of the big wins here, other than the included throw code is smaller, is it prevents all the inlining of `System.Environment::GetResourceString(string)`

Pre
```
Inlines into 06000B7C Lazy`1:.ctor(ref,int):this
  [1 IL=0001 TR=000002 060029B7] [below ALWAYS_INLINE size] Object:.ctor():this
  [0 IL=0014 TR=000068 06002512] [FAILED: unprofitable inline] ArgumentNullException:.ctor(ref):this
  [0 IL=0022 TR=000012 06000B7D] [FAILED: unprofitable inline] Lazy`1:GetObjectFromMode(int):ref
Budget: initialTime=180, finalTime=168, initialBudget=1800, currentBudget=1800
Budget: initialSize=1043, finalSize=1043
Successfully inlined Object:.ctor():this (1 IL bytes) (depth 1) [below ALWAYS_INLINE size]
Successfully inlined Environment:GetResourceString(ref):ref (7 IL bytes) (depth 1) [below ALWAYS_INLINE size]
```
Post
```
Inlines into 06003491 Lazy`1:.ctor(ref,int):this
  [1 IL=0001 TR=000002 06002A4E] [below ALWAYS_INLINE size] Object:.ctor():this
  [0 IL=0011 TR=000061 06000C6F] [FAILED: does not return] ThrowHelper:ThrowArgumentNullException(int)
  [0 IL=0018 TR=000012 06003492] [FAILED: unprofitable inline] Lazy`1:GetObjectFromMode(int):ref
Budget: initialTime=168, finalTime=156, initialBudget=1680, currentBudget=1680
Budget: initialSize=952, finalSize=952
Successfully inlined Object:.ctor():this (1 IL bytes) (depth 1) [below ALWAYS_INLINE size]
```

benaadams commented:
Jitted asm

Pre
```asm
; Assembly listing for method ArraySegment`1:.ctor(ref,int,int):this
; Emitting BLENDED_CODE for X64 CPU with SSE2
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T00] (  5,   5  )   byref  ->  rsi         this
;* V01 TypeCtx      [V01    ] (  0,   0  )    long  ->  zero-ref   
;  V02 arg1         [V02,T01] (  5,   5  )     ref  ->   r8        
;  V03 arg2         [V03,T02] (  5,   5  )     int  ->  rdi        
;  V04 arg3         [V04,T03] (  3,   3  )     int  ->  rbx        
;  V05 loc0         [V05,T04] (  3,   0  )     ref  ->  rsi        
;  V06 tmp1         [V06,T05] (  3,   0  )     ref  ->  rsi        
;  V07 tmp2         [V07,T06] (  3,   0  )     ref  ->  rsi        
;  V08 tmp3         [V08,T07] (  3,   0  )     ref  ->  rsi        
;  V09 tmp4         [V09,T08] (  2,   0  )     ref  ->  rdx        
;  V10 tmp5         [V10,T09] (  2,   0  )     ref  ->  rdi        
;  V11 tmp6         [V11,T10] (  2,   0  )     ref  ->   r8        
;  V12 tmp7         [V12,T11] (  2,   0  )     ref  ->  rdi        
;  V13 tmp8         [V13,T12] (  2,   0  )     ref  ->   r8        
;  V14 tmp9         [V14,T13] (  2,   0  )     ref  ->  rdx        
;  V15 OutArgs      [V15    ] (  1,   1  )  lclBlk (32) [rsp+0x00]  
;
; Lcl frame size = 32
G_M41509_IG01:
       push     rdi
       push     rsi
       push     rbx
       sub      rsp, 32
       mov      rsi, rcx
       mov      edi, r9d
       mov      ebx, dword ptr [rsp+60H]
G_M41509_IG02:
       test     r8, r8
       je       G_M41509_IG08
G_M41509_IG03:
       test     edi, edi
       jl       G_M41509_IG09
G_M41509_IG04:
       test     ebx, ebx
       jl       G_M41509_IG10
G_M41509_IG05:
       mov      edx, dword ptr [r8+8]
       sub      edx, edi
       cmp      edx, ebx
       jl       G_M41509_IG11
G_M41509_IG06:
       mov      rcx, rsi
       mov      rdx, r8
       call     CORINFO_HELP_CHECKED_ASSIGN_REF
       mov      dword ptr [rsi+8], edi
       mov      dword ptr [rsi+12], ebx
G_M41509_IG07:
       add      rsp, 32
       pop      rbx
       pop      rsi
       pop      rdi
       ret      
************** Beginning of cold code **************
G_M41509_IG08:
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       mov      ecx, 0x4911
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      rdx, rax
       mov      rcx, rsi
       call     ArgumentNullException:.ctor(ref):this
       mov      rcx, rsi
       call     CORINFO_HELP_THROW
G_M41509_IG09:
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       mov      ecx, 0x5B7B
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      rdi, rax
       mov      ecx, 0x4C25
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      rcx, rax
       call     Environment:GetResourceStringLocal(ref):ref
       mov      r8, rax
       mov      rdx, rdi
       mov      rcx, rsi
       call     ArgumentOutOfRangeException:.ctor(ref,ref):this
       mov      rcx, rsi
       call     CORINFO_HELP_THROW
G_M41509_IG10:
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       mov      ecx, 0x498F
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      rdi, rax
       mov      ecx, 0x4C25
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      rcx, rax
       call     Environment:GetResourceStringLocal(ref):ref
       mov      r8, rax
       mov      rdx, rdi
       mov      rcx, rsi
       call     ArgumentOutOfRangeException:.ctor(ref,ref):this
       mov      rcx, rsi
       call     CORINFO_HELP_THROW
G_M41509_IG11:
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       mov      ecx, 0x4D75
       call     CORINFO_HELP_STRCNS_CURRENT_MODULE
       mov      rcx, rax
       call     Environment:GetResourceStringLocal(ref):ref
       mov      rdx, rax
       mov      rcx, rsi
       call     ArgumentException:.ctor(ref):this
       mov      rcx, rsi
       call     CORINFO_HELP_THROW
       int3     
; Total bytes of code 314, prolog size 7 for method ArraySegment`1:.ctor(ref,int,int):this
; ============================================================
```
Post
```asm
; Assembly listing for method ArraySegment`1:.ctor(ref,int,int):this
; Emitting BLENDED_CODE for X64 CPU with SSE2
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T00] (  5,   5  )   byref  ->  rsi         this
;* V01 TypeCtx      [V01    ] (  0,   0  )    long  ->  zero-ref   
;  V02 arg1         [V02,T01] (  5,   5  )     ref  ->  rdi        
;  V03 arg2         [V03,T02] (  5,   5  )     int  ->  rbx        
;  V04 arg3         [V04,T03] (  3,   3  )     int  ->  rbp        
;  V05 loc0         [V05    ] (  1,   1  )  lclBlk (32) [rsp+0x00]  
;
; Lcl frame size = 40
G_M41507_IG01:
       push     rdi
       push     rsi
       push     rbp
       push     rbx
       sub      rsp, 40
       mov      rsi, rcx
       mov      rdi, r8
       mov      ebx, r9d
       mov      ebp, dword ptr [rsp+70H]
G_M41507_IG02:
       test     rdi, rdi
       je       G_M41507_IG08
G_M41507_IG03:
       test     ebx, ebx
       jge      SHORT G_M41507_IG04
       mov      ecx, 41
       mov      edx, 4
       call     ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
G_M41507_IG04:
       test     ebp, ebp
       jge      SHORT G_M41507_IG05
       mov      ecx, 16
       mov      edx, 4
       call     ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
G_M41507_IG05:
       mov      ecx, dword ptr [rdi+8]
       sub      ecx, ebx
       cmp      ecx, ebp
       jge      SHORT G_M41507_IG06
       mov      ecx, 23
       call     ThrowHelper:ThrowArgumentException(int)
G_M41507_IG06:
       mov      rcx, rsi
       mov      rdx, rdi
       call     CORINFO_HELP_CHECKED_ASSIGN_REF
       mov      dword ptr [rsi+8], ebx
       mov      dword ptr [rsi+12], ebp
G_M41507_IG07:
       add      rsp, 40
       pop      rbx
       pop      rbp
       pop      rsi
       pop      rdi
       ret      
************** Beginning of cold code **************
G_M41507_IG08:
       mov      ecx, 3
       call     ThrowHelper:ThrowArgumentNullException(int)
       int3     
; Total bytes of code 124, prolog size 8 for method ArraySegment`1:.ctor(ref,int,int):this
; ============================================================
```

benaadams commented:
@AndyAyersMS @mikedn 
The non moved functions are `FAILED: noinline per IL/cached result` rather than `FAILED: does not return`

AndyAyersMS commented:
This is not unexpected -- we can "lose" track of noreturn pretty easily. Some discussion of this in #6095.

mikedn commented:
> The non moved functions are FAILED: noinline per IL/cached result rather than FAILED: does not return

That message is misleading, most of the time a previously inlining attempt failed for a different reason and then all subsequent attempts fail with this "cached result". I looked a bit at `ThrowArgumentException` and it seems that the native code size estimate says that it's not worth inlining it and this estimation is done before we get the "no return" information. I wonder if we couldn't extract "no return" in `fgFindJumpTargets` by counting `ret` and `jmp` instructions.

benaadams commented:
Ah, so il size rather than cached result

ThrowHelper:ThrowArgumentOutOfRangeException(int,int) is (FAILED: noinline per IL/cached result)
```asm
; Assembly listing for method ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
; Emitting BLENDED_CODE for X64 CPU with SSE2
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  3,   2  )     int  ->  rcx        
;  V01 arg1         [V01,T01] (  3,   2  )     int  ->  rsi        
;  V02 tmp0         [V02,T03] (  2,   0  )     ref  ->  rdi        
;  V03 tmp1         [V03,T02] (  3,   0  )     ref  ->  rbx        
;  V04 tmp2         [V04,T04] (  2,   0  )     ref  ->  rsi        
;  V05 tmp3         [V05,T05] (  2,   0  )     ref  ->   r8        
;  V06 OutArgs      [V06    ] (  1,   1  )  lclBlk (32) [rsp+0x00]  
;
; Lcl frame size = 32
G_M43413_IG01:
       push     rdi
       push     rsi
       push     rbx
       sub      rsp, 32
       mov      esi, edx
G_M43413_IG02:
       call     ThrowHelper:GetArgumentName(int):ref
       mov      rdi, rax
       mov      ecx, esi
       call     ThrowHelper:GetResourceName(int):ref
       mov      rsi, rax
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_NEWSFAST
       mov      rbx, rax
       mov      rcx, rsi
       call     Environment:GetResourceStringLocal(ref):ref
       mov      r8, rax
       mov      rdx, rdi
       mov      rcx, rbx
       call     ArgumentOutOfRangeException:.ctor(ref,ref):this
       mov      rcx, rbx
       call     CORINFO_HELP_THROW
       int3     
; Total bytes of code 73, prolog size 7 for method ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
```
whereas
ThrowHelper:ThrowArgumentNullException(int) is (FAILED: does not return)
```asm
; Assembly listing for method ThrowHelper:ThrowArgumentNullException(int)
; Emitting BLENDED_CODE for X64 CPU with SSE2
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  3,   2  )     int  ->  rcx        
;  V01 tmp0         [V01,T01] (  3,   0  )     ref  ->  rdi        
;  V02 tmp1         [V02,T02] (  2,   0  )     ref  ->  rsi        
;  V03 tmp2         [V03,T03] (  2,   0  )     ref  ->  rdx        
;  V04 OutArgs      [V04    ] (  1,   1  )  lclBlk (32) [rsp+0x00]  
;
; Lcl frame size = 40
G_M27416_IG01:
       push     rdi
       push     rsi
       sub      rsp, 40
G_M27416_IG02:
       call     ThrowHelper:GetResourceName(int):ref
       mov      rsi, rax
       lea      rcx, [(reloc)]
       call     CORINFO_HELP_NEWSFAST
       mov      rdi, rax
       mov      rcx, rsi
       call     Environment:GetResourceStringLocal(ref):ref
       mov      rdx, rax
       mov      rcx, rdi
       call     ArgumentNullException:.ctor(ref):this
       mov      rcx, rdi
       call     CORINFO_HELP_THROW
       int3     
; Total bytes of code 57, prolog size 6 for method ThrowHelper:ThrowArgumentNullException(int)
```

benaadams commented:
@mikedn a poor alternative, could add another level of indirection?
```csharp
internal static void ThrowArgumentException(ExceptionResource resource, ExceptionArgument argument) {
    throw GetArgumentException(resource, argument);
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static ArgumentException GetArgumentException(ExceptionResource resource, ExceptionArgument argument) {
    return new ArgumentException(Environment.GetResourceString(GetResourceName(resource)), GetArgumentName(argument));
}
```

mikedn commented:
Ha ha, yeah, that could work. Kind of ugly though.

jkotas commented:
> Let me know if you want anything backed out?

I would do this in generic types and methods only.

> Ha ha, yeah, that could work. Kind of ugly though.

It is the kind of thing that it best to leave to the optimizer, not done manually...

mikedn commented:
> It is the kind of thing that it best to leave to the optimizer, not done manually...

Yup, definitely. And I'm pretty sure that the JIT could handle this, I just need to find some spare time to look at it.

benaadams commented:
> > Let me know if you want anything backed out?

> I would do this in generic types and methods only.

Ok, there are some I'd want to back out then; Filestream especially - its quite hairy in there...

> I'm pretty sure that the JIT could handle this

I'm splitting those 3 methods to test whether it is the size rather than the cache; will report back

benaadams commented:
Splitting them so the initial no return function is short 

Fixes in no inline reason
```
Inlines into 06002A67 ArraySegment`1:.ctor(ref,int,int):this
  [0 IL=0004 TR=000067 06000C7B] [FAILED: does not return] ThrowHelper:ThrowArgumentNullException(int)
  [0 IL=0016 TR=000061 06000C69] [FAILED: does not return] ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
  [0 IL=0028 TR=000054 06000C69] [FAILED: does not return] ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
  [0 IL=0043 TR=000048 06000C77] [FAILED: does not return] ThrowHelper:ThrowArgumentException(int)
Budget: initialTime=270, finalTime=270, initialBudget=2700, currentBudget=2700
Budget: initialSize=1727, finalSize=1727
```
Also the asm cold code
```asm
; Assembly listing for method ArraySegment`1:.ctor(ref,int,int):this
; Emitting BLENDED_CODE for X64 CPU with SSE2
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T00] (  5,   5  )   byref  ->  rsi         this
;* V01 TypeCtx      [V01    ] (  0,   0  )    long  ->  zero-ref   
;  V02 arg1         [V02,T01] (  5,   5  )     ref  ->   r8        
;  V03 arg2         [V03,T02] (  5,   5  )     int  ->  rdi        
;  V04 arg3         [V04,T03] (  3,   3  )     int  ->  rbx        
;  V05 loc0         [V05    ] (  1,   1  )  lclBlk (32) [rsp+0x00]  
;
; Lcl frame size = 32
G_M41510_IG01:
       push     rdi
       push     rsi
       push     rbx
       sub      rsp, 32
       mov      rsi, rcx
       mov      edi, r9d
       mov      ebx, dword ptr [rsp+60H]
G_M41510_IG02:
       test     r8, r8
       je       G_M41510_IG08
G_M41510_IG03:
       test     edi, edi
       jl       G_M41510_IG09
G_M41510_IG04:
       test     ebx, ebx
       jl       G_M41510_IG10
G_M41510_IG05:
       mov      edx, dword ptr [r8+8]
       sub      edx, edi
       cmp      edx, ebx
       jl       G_M41510_IG11
G_M41510_IG06:
       mov      rcx, rsi
       mov      rdx, r8
       call     CORINFO_HELP_CHECKED_ASSIGN_REF
       mov      dword ptr [rsi+8], edi
       mov      dword ptr [rsi+12], ebx
G_M41510_IG07:
       add      rsp, 32
       pop      rbx
       pop      rsi
       pop      rdi
       ret      
************** Beginning of cold code **************
G_M41510_IG08:
       mov      ecx, 3
       call     ThrowHelper:ThrowArgumentNullException(int)
G_M41510_IG09:
       mov      ecx, 41
       mov      edx, 4
       call     ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
G_M41510_IG10:
       mov      ecx, 16
       mov      edx, 4
       call     ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
G_M41510_IG11:
       mov      ecx, 23
       call     ThrowHelper:ThrowArgumentException(int)
       int3     
; Total bytes of code 132, prolog size 7 for method ArraySegment`1:.ctor(ref,int,int):this
```

mikedn commented:
> So it does seem to be the size bail that kicks in first before no return is detected.

Yup. The generated code looks great if this is fixed.

justinvp commented on src/mscorlib/src/System/ThrowHelper.cs:
```diff
@@ -119,323 +189,34 @@ internal static class ThrowHelper {
             }
 
-            return argumentName;
+            return arg.ToString();
```
Could the `Enum.IsDefined` check be removed from release builds? Why not assert as follows?

```c#
internal static string GetArgumentName(ExceptionArgument argument)
{
    Contract.Assert(Enum.IsDefined(typeof(ExceptionArgument), argument),
        "The enum value is not defined, please check the ExceptionArgument Enum.");

    return argument.ToString();
}
```

benaadams replied:
That probably makes more sense :+1:

justinvp commented on src/mscorlib/src/System/ThrowHelper.cs:
```diff
@@ -119,323 +189,24 @@ internal static class ThrowHelper {
-                    break;
+        internal static string GetResourceName(ExceptionResource resource)
+        {
+            Contract.Assert(Enum.IsDefined(typeof(ExceptionArgument), resource),
```
Should be `typeof(ExceptionResource)`.

benaadams replied:
Need something like this https://github.com/dotnet/corefx/issues/10692 ;)

jasonwilliams200OK commented:
test Linux ARM Emulator Cross Debug Build
(#6676; http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/arm_emulator_cross_debug_ubuntu_prtest/1279/)

test Windows_NT x64 Release Priority 1 Build and Test
(#1652; http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/x64_release_windows_nt_pri1_prtest/4507/)

test Windows_NT x86 legacy_backend Checked Build and Test
(similar to #1652, but complaining about `XamlTaskFactory` instead: http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/x86_lb_checked_windows_nt_prtest/2829/)

test Windows_NT x86 ryujit Checked Build and Test
(#6716)

benaadams commented:
@jasonwilliams200OK doesn't seem happy

jasonwilliams200OK commented:
@benaadams, yup, the only difference is this time `Windows_NT x86 ryujit Checked Build and Test` failed with `CodeTaskFactory` instead of `XamlTaskFactory`. I have reported the issue upstream: Microsoft/msbuild#899 (as this is the least we can do, besides submitting a PR with a fix; first need to figure out steps to repro locally with a pretty [SSCCE](http://sscce.org)). :smile:

benaadams commented:
@mikedn @AndyAyersMS could you do something like this instead for return counting? https://github.com/benaadams/coreclr/commit/6de19e73a88b832276be6f78b0286e882b755e1a

I don't know how the Jit flags work though...

benaadams commented:
@mikedn its definitely a `FAILED: unprofitable inline` earlier on
```
**************** Inline Tree
Inlines into 060000BE CleanupWorkList:Add(ref):this
  [0 IL=0007 TR=000004 060036A2] [FAILED: unprofitable inline] List`1:Add(ref):this
Budget: initialTime=99, finalTime=99, initialBudget=990, currentBudget=990
Budget: initialSize=427, finalSize=427
Successfully inlined List`1:get_Count():int:this (7 IL bytes) (depth 1) [below ALWAYS_INLINE size]
Successfully inlined List`1:get_Item(int):ref:this (27 IL bytes) (depth 1) [profitable inline]
Successfully inlined ThrowHelper:ThrowArgumentOutOfRangeException() (10 IL bytes) (depth 2) [below ALWAYS_INLINE size]
Successfully inlined List`1:get_Item(int):ref:this (27 IL bytes) (depth 1) [profitable inline]
Successfully inlined ThrowHelper:ThrowArgumentOutOfRangeException() (10 IL bytes) (depth 2) [below ALWAYS_INLINE size]
**************** Inline Tree
Inlines into 060000BF CleanupWorkList:Destroy():this
  [1 IL=0006 TR=000003 06003697] [below ALWAYS_INLINE size] List`1:get_Count():int:this
  [2 IL=0023 TR=000021 0600369D] [profitable inline] List`1:get_Item(int):ref:this
    [3 IL=0009 TR=000071 06000C69] [below ALWAYS_INLINE size] ThrowHelper:ThrowArgumentOutOfRangeException()
      [0 IL=0004 TR=000080 06000C74] [FAILED: unprofitable inline] ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
  [4 IL=0042 TR=000041 0600369D] [profitable inline] List`1:get_Item(int):ref:this
    [5 IL=0009 TR=000097 06000C69] [below ALWAYS_INLINE size] ThrowHelper:ThrowArgumentOutOfRangeException()
      [0 IL=0004 TR=000106 06000C74] [FAILED: unprofitable inline] ThrowHelper:ThrowArgumentOutOfRangeException(int,int)
  [0 IL=0052 TR=000046 0600008D] [FAILED: noinline per IL/cached result] StubHelpers:SafeHandleRelease(ref)
```
I added `EnhancedLegacyPolicy::DetermineProfitability` as 
```cpp
void EnhancedLegacyPolicy::DetermineProfitability(CORINFO_METHOD_INFO* methodInfo)
{
    if (m_IsNoReturn)
    {
        SetNever(InlineObservation::CALLEE_DOES_NOT_RETURN);
        return;
    }

    LegacyPolicy::DetermineProfitability(methodInfo);
}
```
Then went mostly unconditional on 
```cpp
void EnhancedLegacyPolicy::NoteBool(InlineObservation obs, bool value)
{
    switch (obs)
    {
        case InlineObservation::CALLEE_DOES_NOT_RETURN:
            m_IsNoReturn      = value;
            m_IsNoReturnKnown = true;

            if (value) 
            {
                SetNever(InlineObservation::CALLEE_DOES_NOT_RETURN);
                m_Observation = InlineObservation::CALLEE_DOES_NOT_RETURN;
                m_Decision = InlineDecision::NEVER;
            }
            break;

        default:
            // Pass all other information to the legacy policy
            LegacyPolicy::NoteBool(obs, value);
            break;
    }
}
void EnhancedLegacyPolicy::NoteInt(InlineObservation obs, int value)
{
    if (m_IsNoReturn)
    {
        SetNever(InlineObservation::CALLEE_DOES_NOT_RETURN);
        m_Observation = InlineObservation::CALLEE_DOES_NOT_RETURN;
        m_Decision = InlineDecision::NEVER;
    }
    else
    {
        // Pass all other information to the legacy policy
        LegacyPolicy::NoteInt(obs, value);
    }
}
```
So I think its `Compiler::compCompileHelper` calling `prejitResult.DetermineProfitability(methodInfo);` that marks it before the jit looks at it with Policy?

benaadams commented:
Or rather `void Compiler::fgFindBasicBlocks()` in flowgraph.cpp

benaadams commented:
Non-relocation resolved in https://github.com/dotnet/coreclr/pull/6742

benaadams commented:
With  #6742 drops further (-2.48% vs -1.44%)
```
Total bytes of diff: -77115 (-2.49 % of base)
    diff is an improvement.

Total byte diff includes -29108 bytes from reconciling methods
        Base had  414 unique methods,    29108 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file improvements by size (bytes):
      -77115 : System.Private.CoreLib.dasm (-2.49 % of base)

1 total files with size differences.

Top method regessions by size (bytes):
         642 : System.Private.CoreLib.dasm - Enumerator:System.Collections.IEnumerator.get_Current():ref:this
         624 : System.Private.CoreLib.dasm - ReadOnlyCollection`1:System.Collections.ICollection.CopyTo(ref,int):this
         418 : System.Private.CoreLib.dasm - CustomAttributeData:.ctor(ref,struct):this
         389 : System.Private.CoreLib.dasm - Dictionary`2:System.Collections.ICollection.CopyTo(ref,int):this
         371 : System.Private.CoreLib.dasm - KeyCollection:System.Collections.ICollection.CopyTo(ref,int):this

Top method improvements by size (bytes):
       -3288 : System.Private.CoreLib.dasm - Array:Sort(ref,int,int,ref)
       -2214 : System.Private.CoreLib.dasm - Array:LastIndexOf(ref,struct,int,int):int
       -1540 : System.Private.CoreLib.dasm - Array:IndexOf(ref,struct,int,int):int
       -1311 : System.Private.CoreLib.dasm - __Error:WinIOError(int,ref)
       -1255 : System.Private.CoreLib.dasm - EventProvider:WriteEvent(byref,long,long,ref):bool:this

1341 total methods with size differences.
```

benaadams commented:
With #6742 drops further (-2.49% vs -1.44%)
```
Total bytes of diff: -77277 (-2.49 % of base)
    diff is an improvement.

Total byte diff includes -29108 bytes from reconciling methods
        Base had  414 unique methods,    29108 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file improvements by size (bytes):
      -77277 : System.Private.CoreLib.dasm (-2.49 % of base)

1 total files with size differences.

Top method regessions by size (bytes):
         642 : System.Private.CoreLib.dasm - Enumerator:System.Collections.IEnumerator.get_Current():ref:this
         624 : System.Private.CoreLib.dasm - ReadOnlyCollection`1:System.Collections.ICollection.CopyTo(ref,int):this
         418 : System.Private.CoreLib.dasm - CustomAttributeData:.ctor(ref,struct):this
         389 : System.Private.CoreLib.dasm - Dictionary`2:System.Collections.ICollection.CopyTo(ref,int):this
         371 : System.Private.CoreLib.dasm - KeyCollection:System.Collections.ICollection.CopyTo(ref,int):this

Top method improvements by size (bytes):
       -3288 : System.Private.CoreLib.dasm - Array:Sort(ref,int,int,ref)
       -2214 : System.Private.CoreLib.dasm - Array:LastIndexOf(ref,struct,int,int):int
       -1540 : System.Private.CoreLib.dasm - Array:IndexOf(ref,struct,int,int):int
       -1311 : System.Private.CoreLib.dasm - __Error:WinIOError(int,ref)
       -1255 : System.Private.CoreLib.dasm - EventProvider:WriteEvent(byref,long,long,ref):bool:this

1339 total methods with size differences.
```

benaadams commented:
@dotnet-bot test Linux ARM Emulator Cross Debug Build 
@dotnet-bot test Windows_NT x64 Debug Build and Test 
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test 
@dotnet-bot test Windows_NT x86 ryujit Checked Build and Test

benaadams commented:
@dotnet-bot test Linux ARM Emulator Cross Debug Build 
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test

benaadams commented:
Rebased

benaadams commented:
@dotnet-bot test Windows_NT x86 ryujit Checked Build and Test

benaadams commented:
Closing this as I think most of the big individual ones are done 
* ArraySegment<T> exceptions to ThrowHelper #7058
* Array<T> exceptions to ThrowHelper #7059
* Task<T> Exceptions to ThrowHelper #7076
* ConcurrentDictionary<TKey, TValue> Exceptions to ThrowHelper #7079
* Improve ThrowHelper #7048

The remaining ones in the concurrent collections are in the interface path (hopefully not common use)
